File: Olympiad/COCI/official/2007/contest4_solutions/zbrka.cpp

/*
  Croatian Open Competition in Informatics 2006/2007
  Contest 4 - Task ZBRKA
  Programming language C++
*/

#include <cstdio>

#define maxk 10001
#define mod 1000000007

int n, k, dp[2][maxk]; // za racunanje relacije treba nam samo trenutni i zadnji red

int main(void) {
   scanf( "%d%d", &n, &k );
	dp[1][0] = 1;
	for( int i = 2; i <= n; i++ ) {
		int tmp = 0;
		for( int j = 0; j <= k; j++ ) {
			tmp = (tmp+dp[(i-1)&1][j])%mod;
			if( j >= i ) tmp = (tmp-dp[(i-1)&1][j-i]+mod)%mod;
			dp[i&1][j] = tmp;
		}
	}
	printf( "%d\n", dp[n&1][k] );
	return 0;
}

File: Olympiad/COCI/official/2007/contest4_solutions/zbrka.pas

(*
  Croatian Open Competition in Informatics 2006/2007
  Contest 4 - Task ZBRKA
  Programming language Pascal
*)

program zbrka;

const maxk = 10001;
const mmod = 1000000007;

var
	 sad, prije: integer;
    n, k, i, j, tmp: longint;
    dp: array[0..1,0..maxk] of longint; (* koristimo samo dva retka matrice *)

begin
    readln( n, k );

    dp[1][0] := 1;
    for i := 2 to n do begin
        tmp := 0;
		  prije := (i-1) mod 2;
		  sad := i mod 2;
        for j := 0 to k do begin
            tmp := (tmp + dp[ prije, j ]) mod mmod;
            if( j >= i ) then tmp := (tmp - dp[ prije, j-i ] + mmod ) mod mmod;
            dp[ sad, j ] := tmp;
        end;
    end;

    writeln( dp[ n mod 2, k ] );
end.
